Pokewatch/Program.cs

```diff
@@ -53,74 +60,108 @@ public static void Main(string[] args)
 				Log("[+]Sucessfully signed in to PokemonGo, beginning search.");
 			};
 
-			if (!Search())
-				throw new Exception();
-		}
-
-		private static bool Search()
-		{
-			Queue<FoundPokemon> tweetedPokemon = new Queue<FoundPokemon>();
-			DateTime lastTweet = DateTime.MinValue;
-			Random random = new Random();
-			while (true)
-			{
-				int regionIndex = random.Next(s_config.Regions.Count);
-				if (regionIndex == s_config.Regions.Count)
-					regionIndex = 0;
-
-				Region region = s_config.Regions[regionIndex];
-				Log($"[!]Searching Region: {region.Name}");
-				foreach (Location location in region.Locations)
-				{
-					SetLocation(location);
-
-					//Wait so we don't clobber api and to let the heartbeat catch up to our new location. (Minimum heartbeat time is 4000ms)
-					Thread.Sleep(5000);
-					Log("[!]Searching nearby cells.");
-					RepeatedField<MapCell> mapCells;
-					try
-					{
-						mapCells = s_pogoSession.Map.Cells;
-					}
-					catch
-					{
-						Log("[-]Heartbeat has failed. Terminating Connection.");
-						return false;
-					}
-					foreach (var mapCell in mapCells)
-					{
-						foreach (WildPokemon pokemon in mapCell.WildPokemons)
-						{
-							FoundPokemon foundPokemon = ProcessPokemon(pokemon, tweetedPokemon, lastTweet);
-
-							if (foundPokemon == null)
-								continue;
-
-							string tweet = ComposeTweet(foundPokemon, region);
-
-							try
-							{
-								s_twitterClient.PublishTweet(tweet);
-							}
-							catch(Exception ex)
-							{
-								Log("[-]Tweet failed to publish: " + tweet + " " + ex.Message);
-								continue;
-							}
-
-							Log("[+]Tweet published: " + tweet);
-							lastTweet = DateTime.Now;
-
-							tweetedPokemon.Enqueue(foundPokemon);
-							if (tweetedPokemon.Count > 10)
-								tweetedPokemon.Dequeue();
-						}
-					}
-				}
-				Log("[!]Finished Searching " + region.Name);
-			}
-		}
-
+            s_pogoSession.AccessTokenUpdated += (sender, eventArgs) =>
+            {
+                Log("[+]Access token updated.");
+            };
+
+            s_pogoSession.Player.Inventory.Update += (sender, eventArgs) =>
+            {
+                Log("[+]Inventory was updated.");
+                Console.WriteLine("Inventory was updated.");
+            };
+            s_pogoSession.Map.Update += (sender, eventArgs) =>
+            {
+                Log("[+]Map was updated.");
+                ProcessMap();
+                NextLocation();
+            };
+
+            Console.CancelKeyPress += (sender, eArgs) => {
+                _quitEvent.Set();
+                eArgs.Cancel = true;
+            };
+            
+            _quitEvent.WaitOne();
+        }
+
+	    private static void NextLocation()
+	    {
+	        _locationInt++;
```

Comment:
Are you flattening all locations then iterating down that list rather than grouping by Region? If so, some use of `SelectMany` will probably simplify things, and you shouldn't need two static integers to track it